frontends/common/parseInput.h

```diff
@@ -61,29 +59,33 @@ static const IR::P4Program *parseV1Program(Input &stream, std::string_view sourc
  * on failure. If failure occurs, an error will also be reported.
  */
 template <typename C = P4V1::Converter>
-const IR::P4Program *parseP4File(ParserOptions &options) {
+const IR::P4Program *parseP4File(const ParserOptions &options) {
     BUG_CHECK(&options == &P4CContext::get().options(),
               "Parsing using options that don't match the current "
               "compiler context");
-    FILE *in = nullptr;
+
+    const IR::P4Program *result = nullptr;
     if (options.doNotPreprocess) {
-        in = fopen(options.file.c_str(), "r");
-        if (in == nullptr) {
+        auto *file = fopen(options.file.c_str(), "r");
+        if (file == nullptr) {
             ::error(ErrorType::ERR_NOT_FOUND, "%1%: No such file or directory.", options.file);
             return nullptr;
         }
+        result = options.isv1() ? parseV1Program<FILE *, C>(file, options.file.string(), 1,
```

Comment:
We could probably fix `parseV1Program` to take its first argument by value instead of lvalue reference.  That would require changing the call in parseInput.cpp to use an explicit reference in its call.